Annotation example (+ stack component update)

strickvl opened the pull request:
This is the example for the annotation stack component, including instructions on how to set up the stack(s) for the three support cloud providers.

It also contains some fixes to ensure that the annotation standard steps work together with the multiple cloud providers. These are BREAKING_CHANGES. Specifically, it changes the secret name that is expected for AWS to be able to use this with Label Studio.

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [x] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/advanced-guide/integrations) table and the [corresponding website section](https://zenml.io/integrations).
- [ ] I have added tests to cover my changes.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

fa9r commented on examples/label_studio_annotation/materializers/pillow_image_materializer.py:
```diff
@@ -0,0 +1,53 @@
+    def handle_input(self, data_type: Type[Dict]) -> Dict:
+        """Read from artifact store"""
+        super().handle_input(data_type)
+        temp_dir = tempfile.TemporaryDirectory()
+        io_utils.copy_dir(self.artifact.uri, temp_dir.name)
```
Are we sure this is needed? If so, how come this isn't an issue with other materializers? Or is it?

strickvl replied:
We use this / do this pattern in at least one of our other materializers.

fa9r commented on examples/label_studio_annotation/materializers/pillow_image_materializer.py:
```diff
@@ -0,0 +1,53 @@
+logger = get_logger(__name__)
+
+
+class PillowImageMaterializer(BaseMaterializer):
```
This should probably be turned into a separate Pillow integration. Or perhaps just come as a built-in materializer?
Also, if #812 makes it in, having a materializer for a single PIL image (as we would do in an integration) will be enough.

strickvl replied:
Agreed that we should update and create a new Pillow integration / materializer once #812 is in. I'll create a ticket for it.

fa9r commented on examples/label_studio_annotation/steps/model_deployers.py:
```diff
@@ -0,0 +1,29 @@
+#  permissions and limitations under the License.
+
+from zenml.artifacts.model_artifact import ModelArtifact
+from zenml.integrations.mlflow.steps import mlflow_model_deployer_step
+from zenml.steps import step
+
+model_deployer = mlflow_model_deployer_step()
+
+# TODO: remove below
+
+
```
```suggestion
from zenml.steps import step

```

fa9r commented on examples/label_studio_annotation/steps/prediction_steps.py:
```diff
@@ -0,0 +1,85 @@
+REVERSE_LABEL_MAPPING = {value: key for key, value in LABEL_MAPPING.items()}
+PIPELINE_NAME = "training_pipeline"
+PIPELINE_STEP_NAME = "model_trainer"  # TODO: change to "model_deployer"
+
+
+# @step(enable_cache=False)
+# def prediction_service_loader() -> BaseService:
+#     """Load the model service of our train_evaluate_deploy_pipeline."""
+#     repo = Repository(skip_repository_check=True)
+#     model_deployer = repo.active_stack.model_deployer
+#     services = model_deployer.find_model_server(
+#         pipeline_name=PIPELINE_NAME,
+#         pipeline_step_name=PIPELINE_STEP_NAME,
+#         running=True,
+#     )
+#     service = services[0]
+#     return service
+
+
+# @step
+# def predictor(
+#     service: BaseService,
+#     data: np.ndarray,  # TODO
+# ) -> Output(predictions=list):
+#     """Run a inference request against a prediction service"""
+#     service.start(timeout=10)  # should be a NOP if already started
+#     # TODO: iterate over images, convert predictions to label studio format
+#     prediction = service.predict(data)
+#     prediction = prediction.argmax(axis=-1)
+#     print(f"Prediction is: {[prediction.tolist()]}")
+#     return [prediction.tolist()]
```
```suggestion
```

fa9r commented on examples/label_studio_annotation/steps/pytorch_trainer.py:
```diff
@@ -0,0 +1,267 @@
+    return model
+
+
+class CustomDataset:
```
Could use some docstrings

fa9r commented on examples/label_studio_annotation/steps/pytorch_trainer.py:
```diff
@@ -0,0 +1,267 @@
+
+
+@step(enable_cache=False)
+def pytorch_model_trainer(
```
Could use some docstring

fa9r commented on examples/label_studio_annotation/README.md:
```diff
@@ -0,0 +1,318 @@
+training and annotation into a loop) with Label Studio. This uses a combination
+of user-defined steps as well as some built-in steps that ZenML provides.
+
+## Basic guide to running this example
+
+This example requires a few more steps than usual, given that it requires a
+cloud artifact store. Instructions for Azure, AWS S3 and GCP can be found in the
+following guide.
+
+## 🗺 Overview
+
+NOTE: This example currently only runs with a cloud artifact store. [This guide
+by Label Studio](https://labelstud.io/guide/storage.html) contains more details
+if you have trouble understanding the instructions below.
```
This feels a bit redundant, would propose to remove the second paragraph and move the first one down:

```suggestion
## 🗺 Overview

This example requires a few more steps than usual, given that it requires a
cloud artifact store. Instructions for Azure, AWS S3 and GCP can be found in the
following guide.
```

fa9r commented on examples/label_studio_annotation/README.md:
```diff
@@ -0,0 +1,318 @@
+# 🖥 Run the example
+
+## 👣 Step-by-Step
+
```
I'd suggest to add a header "Infrastructure Setup" here under which you group the three sections for AWS, Azure, GCP as subheaders and also add a short comment in the beginning clarifying again that you can use any of the three cloud providers. 

(When I first saw this it felt like you need to do all three setups since "Step-by-Step" sounds to me like you need to follow the subsequent steps one-by-one.)

fa9r commented on examples/label_studio_annotation/README.md:
````diff
@@ -0,0 +1,318 @@
+
+### 🥞 Set up your stack for Microsoft Azure
+
+In order to run this example, you need to install and initialize ZenML and Label
+Studio.
+
+```shell
+zenml integration install label_studio pytorch azure
+
+# pull example
+zenml example pull label_studio_annotation
+cd zenml_examples/label_studio_annotation
+
+# Initialize ZenML repo
+zenml init
+```
````
This is essentially the same for all three cloud providers, only `zenml integration install azure` is provider-specific here. I would recommend to move the rest into a separate section "ZenML Setup" or similar to remove code duplication and make the README a bit shorter

strickvl replied:
Yep I moved that now.

fa9r commented on examples/label_studio_annotation/README.md:
````diff
@@ -0,0 +1,318 @@
+
+You will next need to obtain your Label Studio API key. This will give you access to the web annotation interface.
+
+```shell
+# start a temporary / one-off label-studio instance to get your API key
+label-studio start -p 8094
+```
+
+First order of business is to sign up if you haven't done so already, using an email and a password. Then visit [http://localhost:8094/](http://localhost:8094/) to log in, and then visit [http://localhost:8094/user/account](http://localhost:8094/user/account) and get your Label Studio API key (from the upper right hand corner). You will need it for the next step. `Ctrl-c` out of the Label Studio  ...
+
+```shell
+zenml secret register <LABEL_STUDIO_SECRET_NAME> --api_key="<YOUR_API_KEY>"
+
+zenml annotator register <YOUR_LABEL_STUDIO_ANNOTATOR> --flavor label_studio --authentication_secret="<LABEL_STUDIO_SECRET_NAME>"
+
+zenml stack update <ANNOTATION_STACK_NAME> -an <YOUR_LABEL_STUDIO_ANNOTATOR>
+
+zenml stack up
+```
````
This is also not cloud provider-specific. Similar to above, I would suggest to move this to a separate section "Label Studio Setup" or similar instead of having it duplicated 3x.

strickvl replied:
Also has been moved, but the stack setup steps can't be moved elsewhere unfortunately as there are some order dependencies that are a bit fiddly.

schustmi commented on examples/label_studio_annotation/README.md:
```diff
@@ -0,0 +1,318 @@
+by Label Studio](https://labelstud.io/guide/storage.html) contains more details
+if you have trouble understanding the instructions below.
+
+For a full video walkthrough on how to run this email, please check out [our
```
```suggestion
For a full video walkthrough on how to run this example, please check out [our
```

fa9r commented on examples/label_studio_annotation/README.md:
```diff
@@ -0,0 +1,318 @@
+### ▶️ Run the Code
+
+
+This will initialize the daemon server which Label Studio requires, albeit running on a default port of 8093.
```
This line should be moved after `zenml stack up` I guess?

schustmi replied:
Should this sentence be part of `Run the Code`?

schustmi commented on examples/label_studio_annotation/README.md:
````diff
@@ -0,0 +1,318 @@
+zenml stack update <YOUR_AWS_ZENML_STACK_NAME> -a <YOUR_CLOUD_ARTIFACT_STORE>
+```
+
+You will next need to obtain your Label Studio API key. This will give you access to the web annotation interface.
````
Can we move this up ahead of the split into the three cloud providers to reduce the duplication?

fa9r commented on examples/label_studio_annotation/README.md:
```diff
@@ -0,0 +1,318 @@
+
+This will initialize the daemon server which Label Studio requires, albeit running on a default port of 8093.
+
+Run the pipeline with:
```
This needs a bit more explanation, e.g.:

#### 1. Run the training pipeline
This does ...
You should see ...

#### 2. Run the inference pipeline
...
This creates new data automatically written to label studio backend blabla
...

#### 3. Open Label Studio and annotate the new data
Go to this URL, click here, do that, now you should see such and such
....

#### 4. Rerun the training data
This is what happens, trains on new data, ...

#### 5. Rerun the inference pipeline
...

#### (Optional) 6. Annotate new data in label studio again
...

schustmi commented on examples/label_studio_annotation/README.md:
```diff
@@ -0,0 +1,318 @@
+### 🧽 Clean up
+
+In order to clean up, delete any relevant cloud infrastructure that was created
+above (the S3 bucket is the main one, along with any permissions) and the
```
not just s3 but also azure/gcp + secrets manager right?

schustmi commented on examples/label_studio_annotation/steps/deployment_triggers.py:
```diff
@@ -0,0 +1,6 @@
+
+@step
+def deployment_trigger() -> bool:
+    return True  # TODO: only return True if dataset changed
```
Remove the TODO?

schustmi commented on examples/label_studio_annotation/steps/load_image_data_step.py:
```diff
@@ -0,0 +1,45 @@
+
+
+class LoadImageDataConfig(BaseStepConfig):
+    base_path = os.path.join(os.getcwd(), "data")
```
I'd not make it relative to the current working directory, as that might be dynamic. Instead relative to this file is safer IMO

strickvl replied:
This seems to be a philosophical difference between you and @fa9r. I'll make it a relative link.

fa9r commented on examples/label_studio_annotation/README.md:
````diff
@@ -0,0 +1,318 @@
+python run.py aws --inference --rerun
+```
+
+## CLI Commands for the Label Studio Integration
````
Instead of listing what commands there are I'd recommend to show it at the respective place in the previous section, e.g., if you follow the structure I proposed in the previous comment then you could under 2. do `zenml annotator dataset list` at the end to show that there is now something in label studio, and then under 3. have the `zenml annotator dataset ... stats` command to show that X images need to be labeled now and then run `zenml annotator dataset ... annotate` to open the UI and do the annotation. Screenshots of these steps (label studio UI, stats CLI output, ...) would also be nice to include here so users can understand the example without having to actually run it.

schustmi commented on examples/label_studio_annotation/steps/model_deployers.py:
```diff
@@ -0,0 +1,29 @@
+
+model_deployer = mlflow_model_deployer_step()
+
+# TODO: remove below
```
Remove?

schustmi commented on examples/label_studio_annotation/steps/prediction_steps.py:
```diff
@@ -0,0 +1,85 @@
+
+REVERSE_LABEL_MAPPING = {value: key for key, value in LABEL_MAPPING.items()}
+PIPELINE_NAME = "training_pipeline"
+PIPELINE_STEP_NAME = "model_trainer"  # TODO: change to "model_deployer"
```
Many todos and commented out code in this file

schustmi commented on examples/label_studio_annotation/steps/pytorch_trainer.py:
```diff
@@ -0,0 +1,267 @@
+
+LABEL_MAPPING = {label: idx for idx, label in enumerate(LABELS)}
+
+PIPELINE_NAME = "training_pipeline"  # TODO: cleanup
```
remove?

schustmi commented on src/zenml/integrations/label_studio/label_studio_utils.py:
```diff
@@ -118,3 +104,51 @@ def convert_pred_filenames_to_task_ids(
     ]
+
+
+def is_s3_url(url: str) -> bool:
```
Are these used anywhere?

strickvl replied:
Yes. Esp now.

fa9r commented on examples/label_studio_annotation/README.md:
```diff
@@ -0,0 +1,318 @@
+
+### 🧽 Clean up
+
+In order to clean up, delete any relevant cloud infrastructure that was created
+above (the S3 bucket is the main one, along with any permissions) and the
+remaining ZenML references.
```
Would be cool to have a short list of what exactly needs to be deleted and/or links to guides how to do it. @wjayesh  helped me with this for the k8s orchestrator example/blog, maybe we could copy-paste that to here (for AWS at least).

fa9r commented on examples/label_studio_annotation/README.md:
```diff
@@ -0,0 +1,318 @@
+own annotation tool in ZenML
+check out our [docs](https://docs.zenml.io/mlops-stacks/annotators).
+
+We also have extensive CLI docs for the
+[annotator](https://apidocs.zenml.io/latest/api_docs/annotators/) stack
+component.
```
```suggestion
To explore other available CLI commands related to the annotator stack components, check out the
[CLI docs annotator section](https://apidocs.zenml.io/latest/api_docs/repository/#zenml.repository.Repository).
```

fa9r reviewed (approved):
Awesome, looking forward to get this live! 🤜 🤛 

The two major suggestions I still have are:
- reduce redundancy in the README
- explain the actual "Run the Code" section in more detail. 

Otherwise looks good to go!

strickvl commented:
@schustmi @fa9r updated and addressed all your comments, I believe. (Am off for next two days, so wanted this ready to merge at least, and not requiring you to do anything for it).
